blockingconcurrentqueue.h

```diff
@@ -897,29 +1283,69 @@ class BlockingConcurrentQueue
 	// and at most max.
 	// Using a negative timeout indicates an indefinite timeout,
 	// and is thus functionally equivalent to calling wait_dequeue_bulk.
+  // Also returns 0 if the queue is finished (closed and drained of all remaining elements).
 	// Never allocates. Thread-safe.
 	template<typename It>
 	inline size_t wait_dequeue_bulk_timed(consumer_token_t& token, It itemFirst, size_t max, std::int64_t timeout_usecs)
 	{
 		size_t count = 0;
-		max = (size_t)sema->waitMany((LightweightSemaphore::ssize_t)(ssize_t)max, timeout_usecs);
+		max = (size_t)sema_dequeue->waitMany((LightweightSemaphore::ssize_t)(ssize_t)max, timeout_usecs);
 		while (count != max) {
-			count += inner.template try_dequeue_bulk<It&>(token, itemFirst, max - count);
+			size_t got = inner.template try_dequeue_bulk<It&>(token, itemFirst, max - count);
+      if (got == 0 && closed()) {
+        sema_dequeue->signal();
+        return count;
+      }
+      count += got;
 		}
+		sema_enqueue->signal((LightweightSemaphore::ssize_t)(ssize_t)count);
 		return count;
 	}
 	
 	// Attempts to dequeue several elements from the queue using an explicit consumer token.
 	// Returns the number of items actually dequeued, which can
 	// be 0 if the timeout expires while waiting for elements,
 	// and at most max.
+  // Also returns 0 if the queue is finished (closed and drained of all remaining elements).
 	// Never allocates. Thread-safe.
 	template<typename It, typename Rep, typename Period>
 	inline size_t wait_dequeue_bulk_timed(consumer_token_t& token, It itemFirst, size_t max, std::chrono::duration<Rep, Period> const& timeout)
     {
         return wait_dequeue_bulk_timed<It&>(token, itemFirst, max, std::chrono::duration_cast<std::chrono::microseconds>(timeout).count());
     }
-	
+
+  // Close the queue.
+  // This signals that no more data will be enqueued to dequeue operations.
+  // Further dequeue operations will behave normally until the queue is
+  // completely drained (finished() == true). After that, any dequeue calls
+  // will immediately fail and return.
```

Comment:
How? The semaphore says an extra item is available; the `wait_dequeue` methods will do the wrong thing, and the plain `dequeue` methods don't check the flag...

Reply:
My idea here was to use a sort of virtual sentinel element.

```
while (!inner.try_dequeue(item)) {
  // This is supposed to check for the sentinel;
  // there is an element 'in' the semaphore, all
  // enqueue ops *must* have stabilized (because
  // close() demands so), so  the only case where
  // dequeue could fail, is if the inner queue is
  // truly depleted.
  if (closed()) {
    // Add a new sentinel element, so the same logic
    // will work again if wait_dequeue is called again.
    sema_dequeue->signal();
    return false;
  }
  ...
}
```


And you are right, I completely forgot to update the logic in
the old dequeue functions!